### PR TITLE
Fix #691 SingleColumnValueFilter NOT_EQUALS empty value not working,

### LINKED ISSUE #691: SingleColumnValueFilter NOT_EQUALS empty value not working, rows are selected
Hello,

We face an issue when filtering the string values not equals to "", they are selected.

In order to solve this issue, we can set superior (> "")
I don't see the explanation for this issue, seems that empty value is specific.

Moreover (this is not a bug) I think we can not remove cells, so we had the idea to put the values to empty, for longs, etc.
But these rows are selected with the inferior filter, they are selected. So for the INF & STRICT_INF, we must add STRICT_SUP empty value.
Do you have a better solution, or can we delete a cell, without deleting and inserting again the row (we potentially have a concurrent effect on selecting values)

Thank you

### PR DESCRIPTION
**Pull request by ajaaym**
Fix #691 SingleColumnValueFilter NOT_EQUALS empty value not working, rows are selected

### CODE REVIEW

File: bigtable-client-core-parent/bigtable-hbase-integration-tests-common/src/test/java/com/google/cloud/bigtable/hbase/AbstractTestFilters.java [hunk 1]
```diff
@@ -1547,6 +1547,51 @@ public void testSingleColumnValueFilter() throws IOException {
...
+    // This is not intuitive. In order to get filter.setLatestVersionOnly to have an effect,
+    // we must enable the scanner to see more versions:
+    scanAddVersion(scan,3);
+    ByteArrayComparable valueComparable1 = new BinaryComparator(value1_2);
+    SingleColumnValueFilter filter =
+        new SingleColumnValueFilter(COLUMN_FAMILY, qualifier1, CompareOp.NOT_EQUAL, valueComparable1);
+    filter.setFilterIfMissing(false);
+    filter.setLatestVersionOnly(true);
+
+    // a: Qualifier exists in the row and the value NOT_EQUAL to EMPTY_STRING (row1)
```
[the remark this answers is not in the document]
Comment (ajaaym) [reply bigtable-client-core-parent/bigtable-hbase-integration-tests-common/src/test/java/com/google/cloud/bigtable/hbase/AbstractTestFilters.java]: its at line 1584, do you want me to name valueComparable1 to EMPTY_STRING comparator?

File: bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/adapters/filters/ValueFilterAdapter.java [hunk 2]
```diff
@@ -88,10 +88,13 @@ private Filter adaptBinaryComparator(
           return range().startClosed(value).endClosed(value);
         }
       case NOT_EQUAL:
-        // This strictly less than + strictly greater than:
-        return FILTERS.interleave()
-            .filter(range().endOpen(value))
-            .filter(range().startOpen(value));
+        if(comparator.getValue().length == 0) {
+          //Special case for NOT_EQUAL to EMPTY_STRING
+          return FILTERS.value().regex(".+");
```
[the remark this answers is not in the document]
Comment (ajaaym) [reply bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/adapters/filters/ValueFilterAdapter.java]: That doesnt work. It matches with empty bytes too.

**Comment by ajaaym** [comment]
@sduskis PTAL

File: bigtable-client-core-parent/bigtable-hbase-integration-tests-common/src/test/java/com/google/cloud/bigtable/hbase/AbstractTestFilters.java [hunk 3]
```diff
@@ -1547,6 +1547,52 @@ public void testSingleColumnValueFilter() throws IOException {
...
+    byte[] emptyStringValue = "".getBytes();
+
+
+    Table table = getDefaultTable();
+    Put put = new Put(rowKey1);
+    put.addColumn(COLUMN_FAMILY, qualifier1, nonEmptyStringValue);
+    table.put(put);
+
+    put = new Put(rowKey2);
+    put.addColumn(COLUMN_FAMILY, qualifier1, emptyStringValue);
```
Comment (sduskis) [inline bigtable-client-core-parent/bigtable-hbase-integration-tests-common/src/test/java/com/google/cloud/bigtable/hbase/AbstractTestFilters.java]: Empty value == null for Cloud Bigtable.  For rowKey2, I would think that `table.get(new Get(rowKey2))` will never return any values.

File: bigtable-client-core-parent/bigtable-hbase-integration-tests-common/src/test/java/com/google/cloud/bigtable/hbase/AbstractTestFilters.java [hunk 4]
```diff
@@ -1547,6 +1547,52 @@ public void testSingleColumnValueFilter() throws IOException {
...
+    table.put(put);
+
+    Result[] results;
+    Scan scan = new Scan();
+    scan.addColumn(COLUMN_FAMILY, qualifier1);
```
Comment (sduskis) [inline bigtable-client-core-parent/bigtable-hbase-integration-tests-common/src/test/java/com/google/cloud/bigtable/hbase/AbstractTestFilters.java]: please set a prefix on the scan for only "scvfrk"

File: bigtable-client-core-parent/bigtable-hbase-integration-tests-common/src/test/java/com/google/cloud/bigtable/hbase/AbstractTestFilters.java (hunk 3 above)
Comment (sduskis) [reply bigtable-client-core-parent/bigtable-hbase-integration-tests-common/src/test/java/com/google/cloud/bigtable/hbase/AbstractTestFilters.java]: Please add a second qualifier on both puts to make sure that we don't return rowKey2

Comment (ajaaym) [reply bigtable-client-core-parent/bigtable-hbase-integration-tests-common/src/test/java/com/google/cloud/bigtable/hbase/AbstractTestFilters.java]: Table.get(new Get(rowKey2)) does return the row. 

Non empty value for second qualifier will return the rowKey2 if we put filter of NOT_EQUAL to empty string in cell value.

File: bigtable-client-core-parent/bigtable-hbase-integration-tests-common/src/test/java/com/google/cloud/bigtable/hbase/AbstractTestFilters.java (hunk 4 above)
Comment (ajaaym) [reply bigtable-client-core-parent/bigtable-hbase-integration-tests-common/src/test/java/com/google/cloud/bigtable/hbase/AbstractTestFilters.java]: I just tried this, result is same. Row with rowKey2 is not coming. Dont see use of prefix filter.

File: bigtable-client-core-parent/bigtable-hbase-integration-tests-common/src/test/java/com/google/cloud/bigtable/hbase/AbstractTestFilters.java (hunk 3 above)
Comment (ajaaym) [reply bigtable-client-core-parent/bigtable-hbase-integration-tests-common/src/test/java/com/google/cloud/bigtable/hbase/AbstractTestFilters.java]: I just tried adding second qualifier in both row with some string bytes in value, it returns both row with cells which has value(filtering out cell with empty value), using SingleColumnFIlter with NOT_EQUAL and empty string/value.

File: bigtable-client-core-parent/bigtable-hbase-integration-tests-common/src/test/java/com/google/cloud/bigtable/hbase/AbstractTestFilters.java (hunk 4 above)
Comment (sduskis) [reply bigtable-client-core-parent/bigtable-hbase-integration-tests-common/src/test/java/com/google/cloud/bigtable/hbase/AbstractTestFilters.java]: Please use `scan.setRowPrefixFilter(ByteUtil.toBytes("scvfrk"))` for the prefix.

Comment (ajaaym) [reply bigtable-client-core-parent/bigtable-hbase-integration-tests-common/src/test/java/com/google/cloud/bigtable/hbase/AbstractTestFilters.java]: done

File: bigtable-client-core-parent/bigtable-hbase-integration-tests-common/src/test/java/com/google/cloud/bigtable/hbase/AbstractTestFilters.java [hunk 5]
```diff
@@ -1547,6 +1547,55 @@ public void testSingleColumnValueFilter() throws IOException {
...
+    scan.setRowPrefixFilter(Bytes.toBytes("scvfrk"));
+
+    // This is not intuitive. In order to get filter.setLatestVersionOnly to have an effect,
+    // we must enable the scanner to see more versions:
+    scanAddVersion(scan, 3);
+    ByteArrayComparable emptyStringValueComparable = new BinaryComparator(emptyStringValue);
+    SingleColumnValueFilter filter =
+        new SingleColumnValueFilter(COLUMN_FAMILY, qualifier2, CompareOp.NOT_EQUAL, emptyStringValueComparable);
+    filter.setFilterIfMissing(false);
```
Comment (sduskis) [inline bigtable-client-core-parent/bigtable-hbase-integration-tests-common/src/test/java/com/google/cloud/bigtable/hbase/AbstractTestFilters.java]: Please remove `filter.setFilterIfMissing(false);` and `filter.setLatestVersionOnly(true);` which are the default settings.

**Comment by ajaaym** [comment]
closing this merged other commit by mistake will open new one.

**Comment by mbrukman** [comment]
@ajaaym — note that you don't have to abandon a PR if you accidentally merged extra commits. You can simply fix your local branch and then `git push -f` to overwrite the remote branch in-place, even if they have incompatible commits (e.g., you rewrite history to remove commits or undo a merge).

**Pull request closed without merging**